Add socket CAN dispatcher support

jukkar opened the pull request:
This allows application(s) to listen same CAN-IDs for the same network interface.

jukkar committed:
net: sockets: can: Add dispatcher

We need to dispatch the received CAN frame if there are multiple
sockets interested in the same CAN-IDs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: can: Add second RX socket receiver

This is done so that we can test socket CAN dispatcher.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

alexanderwachter commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -82,32 +93,76 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
-	/* Normal packet */
-	net_pkt_set_eof(pkt, false);
+		/* if pkt is NULL, EOF */
+		if (!clone) {
```
Doesn't exist an error code for out of memory?

jukkar replied:
Yes, that needs to be checked.

pfalcon commented:
This needs rebase.

jukkar committed:
net: context: Set local address properly for AF_CAN connections

At the moment there is no real address for local CANBUS socket,
but we can still set protocol family of local socket to AF_CAN
so that for example net-shell "net conn" command can show
information about it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print connection information properly for AF_CAN

CANBUS socket information was just printing unknown information.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -377,6 +498,24 @@ static int can_sock_setsockopt_vmeth(void *obj, int level, int optname,
+			}
+		}
+
 		return api->setsockopt(dev, obj, level, optname, optval,
```
Good point, I will add that.

jukkar commented:
New version:
* Built on top of #14193 and #14201
* Tested on nucleo_l432kc

jukkar commented:
@alexanderwachter can you have a re-review?

alexanderwachter commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -82,32 +93,82 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
+			continue;
+		}
+
+		if (pkt != NULL && ARRAY_SIZE(receivers) > 1) {
```
Don't you need to check for i=0 instead of ARRAY_SIZE(receivers) > 1 here?
Because the original pkt must can be passed so it is unrfed somewhere.

jukkar replied:
> Don't you need to check for i != 0 instead of ARRAY_SIZE(receivers) > 1 here?

In order to simplify pkt housekeeping, I just cloned the packet if there are multiple receiver. After revisiting this one after some time, it looks quite non-optimal. I will rework this a bit more.

alexanderwachter replied:
Still I would keep the original pkt and only clone if i > 0.
But this works too.

alexanderwachter commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -398,13 +541,46 @@ static int can_sock_setsockopt_vmeth(void *obj, int level, int optname,
+					&((struct can_filter *)optval)[i],
+					&zfilter);
+
+				ret = api->setsockopt(dev, obj, level, optname,
```
You should check here if someone else already attached the same filter. If yes, there is no need to attach it again.

alexanderwachter commented:
When you send a message you should check the registered filters and if the id matches, the message should be cloned to this receiver.
What about detaching the filters after the socket is closed? I can't find this in the actual implementation.

jukkar commented:
Lets target this for 1.15

jukkar commented:
Rebased against latest master

jukkar commented:
This still needs some TLC as @alexanderwachter comments needs to be addressed. I am working on it but the progress is slow because of other pending tasks.

jukkar commented:
Fixed merge conflicts, no other changes yet.

jukkar committed:
net: connection: The connection logic was not init properly

We need to initialize the connection.c for UDP, TCP, PACKET socket
and CANBUS sockets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated according to comments.

tbursztyka commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -77,32 +88,87 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
-		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+	for (i = 0; i < ARRAY_SIZE(receivers); i++) {
+		struct zcan_frame *zframe = (struct zcan_frame *)
+							pkt->buffer->data;
```
you could use net_pkt_data() instead

tbursztyka commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -361,13 +425,113 @@ static int can_sock_getsockopt_vmeth(void *obj, int level, int optname,
 		const struct canbus_api *api;
 		struct net_if *iface;
 		struct device *dev;
+		int ret;
```
I would take the opportunity to refactor this function to avoid too much indentation.

```
if (level != SOL_CAR_RAW) {
        return zcan_setsockopt_ctx(obj, level, optname, optval, optlen);
}

... 
```

jukkar commented:
Updated according to comments.

jukkar commented:
@alexanderwachter can you review this, all the comments you had should be ok now.

alexanderwachter commented:
There is still no filter detaching when the socket is closed.
The filter is still attached if an application for example opens a socket, set the filter socket opt and then closes the socket. If the application does this several times we would run out of filters very quickly. Another problem with leaving the filter attaches is that this filter causes interrupts even nobody is waiting for this frame.

jukkar committed:
net: sockets: can: Close the socket cleanly

If the socket is closed, then do CAN detach if that is needed.
This way the CAN interrupts are not received if there are no
CAN sockets listening the data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: socket: can: Close the socket periodically

This is done only for testing purposes, in real life the socket
would be closed if it is not used or needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed the socket closing (detach filters if needed).
